main/src/ca/uwaterloo/flix/language/phase/ConstraintResolution.scala

```diff
@@ -0,0 +1,716 @@
+    case KindedAst.Sig(_, _, None) => Validation.success(Substitution.empty)
+    case KindedAst.Sig(_, spec, Some(_)) => visitSpec(spec, infResult, renv0, tconstrs0, cenv0, eqEnv0, root)
+  }
+
+  /**
+    * Resolves constraints in the given spec using the given inference result.
+    */
+  def visitSpec(spec: KindedAst.Spec, infResult: InfResult, renv0: RigidityEnv, tconstrs0: List[Ast.TypeConstraint], cenv0: Map[Symbol.ClassSym, Ast.ClassContext], eqEnv0: ListMap[Symbol.AssocTypeSym, Ast.AssocTypeDef], root: KindedAst.Root)(implicit flix: Flix): Validation[Substitution, TypeError] = spec match {
+    case KindedAst.Spec(_, _, _, _, fparams, _, tpe, eff, tconstrs, econstrs, loc) =>
+
+      val InfResult(infConstrs, infTpe, infEff, infRenv) = infResult
+
+      // The initial substitution maps from formal parameters to their types
+      val initialSubst = fparams.foldLeft(Substitution.empty) {
+        case (acc, KindedAst.FormalParam(sym, mod, tpe, src, loc)) => acc ++ Substitution.singleton(sym.tvar.sym, openOuterSchema(tpe)(Level.Top, flix))
+      }
+
+      // Wildcard tparams are not counted in the tparams, so we need to traverse the types to get them.
+      val allTparams = tpe.typeVars ++ eff.typeVars ++ fparams.flatMap(_.tpe.typeVars) ++ econstrs.flatMap(_.tpe2.typeVars)
+
+      // The rigidity environment is made up of:
+      // 1. rigid variables from the context (e.g. from instance type parameters)
+      // 2. rigid variables from type inference (e.g. regions)
+      // 3. rigid variables from declared function type parameters
+      val renv = allTparams.foldLeft(infRenv ++ renv0) {
+        case (acc, Type.Var(sym, _)) => acc.markRigid(sym)
+      }
+
+      // The class and equality environments are made up of:
+      // 1. constraints from the context (e.g. constraints on instances and classes, plus global constraints)
+      // 2. constraints from the function signature
+      val cenv = expandClassEnv(cenv0, tconstrs ++ tconstrs0)
+      val eqEnv = expandEqualityEnv(eqEnv0, econstrs) // TODO ASSOC-TYPES allow econstrs on instances
+
+      // We add extra constraints for the declared type and effect
+      val tpeConstr = TypingConstraint.Equality(tpe, infTpe, Provenance.ExpectType(expected = tpe, actual = infTpe, loc))
+      val effConstr = TypingConstraint.Equality(eff, infEff, Provenance.ExpectEffect(expected = eff, actual = infEff, loc))
+      val constrs = tpeConstr :: effConstr :: infConstrs
+
+
+      resolve(constrs, initialSubst, renv, cenv, eqEnv).flatMap {
+        case ResolutionResult(subst, deferred, _) =>
+          Debug.stopLogging()
+
+          // If there are any constraints we could not resolve, then we report an error.
+          // TODO ASSOC-TYPES here we only consider the first error
+          deferred match {
```

Comment:
You say deferred, but I say pending or unresolved. I guess we should harmonize that, but probably not important now.